src/tools/clippy/clippy_lints/src/ptr.rs

```diff
@@ -460,13 +460,25 @@ fn check_fn_args<'cx, 'tcx: 'cx>(
                             }
                             None
                         }) {
+                            let binding_item = match lifetime.res {
+                                LifetimeName::Param(param_def_id) => Some(cx.tcx.parent(param_def_id.to_def_id())),
+                                _ => None,
+                            };
+                            
                             if !lifetime.is_anonymous()
                                 && fn_sig
                                     .output()
                                     .walk()
                                     .filter_map(|arg| {
                                         arg.as_region().and_then(|lifetime| match lifetime.kind() {
-                                            ty::ReEarlyParam(r) => Some(r.def_id),
+                                            ty::ReEarlyParam(r) if binding_item.is_some() => Some(
+                                                cx.tcx.generics_of(binding_item.unwrap()).region_param(r, cx.tcx).def_id
+                                            ),
+                                            // It doesn't seem entirely correct to return none here, since all early-bound
+                                            // parameters will have a corresponding `DefId`. It shouldn't matter though as if
+                                            // `binding_item` is `None` then the `any` afterwards is going to return `false`
+                                            // no matter what `DefId`'s we do or do not return. 
```

Comment:
This is a good catch actually. Should this entire `if` be rewritten to:
```rust
if let LifetimeName::Param(param_def_id) = lifetime.res
    && !lifetime.is_anonymous()
    && fn_sig...
        .any(|def_id| def_id.as_local().is_some_and(|def_id| def_id == param_def_id))
```

This would simplify this code quite a bit.